checker/src/org/checkerframework/checker/lock/wholefiles/LockVisitor.java

```diff
@@ -0,0 +1,580 @@
+                    if (AnnotationUtils.hasElementValue(annotationMirror, "value")) {
+                        List<String> guardedByValue = AnnotationUtils.getElementValueArray(annotationMirror, "value", String.class, false);
+
+                        for(String lockExpression : guardedByValue) {
+                            if (translateItselfToThis && lockExpression.equals("itself")) {
+                                lockExpression = "this";
+                            }
+                            preconditions.add(Pair.of(lockExpression, checkerLockHeldClass.toString().substring(10 /* "interface " */)));
+                        }
+                    }
+                }
+            }
+        }
+
+        return preconditions;
+    }
+
+    @Override
+    protected void commonAssignmentCheck(AnnotatedTypeMirror varType,
+            AnnotatedTypeMirror valueType, Tree valueTree, /*@CompilerMessageKey*/ String errorKey,
+            boolean isLocalVariableAssignement) {
+
+        if (valueType.getKind() == TypeKind.NULL) {
+            // Avoid issuing warnings about 'null' not matching the type of the variable.
+            return;
+        }
+
+        Kind valueTreeKind = valueTree.getKind();
+
+        switch(valueTreeKind) {
+            case NEW_CLASS:
+            case NEW_ARRAY:
+                // Avoid issuing warnings for: @GuardedBy(<something>) Object o = new Object();
+                // Do NOT do this if the LHS is @GuardedByBottom.
+                if (!AnnotationUtils.areSameIgnoringValues(varType.getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE), GUARDEDBYBOTTOM))
+                    return;
+                break;
+            case INT_LITERAL:
+            case LONG_LITERAL:
+            case FLOAT_LITERAL:
+            case DOUBLE_LITERAL:
+            case BOOLEAN_LITERAL:
+            case CHAR_LITERAL:
+            case STRING_LITERAL:
+                // Avoid issuing warnings for: @GuardedBy(<something>) Object o; o = <some literal>;
+                // Do NOT do this if the LHS is @GuardedByBottom.
+                if (!AnnotationUtils.areSameIgnoringValues(varType.getAnnotationInHierarchy(GUARDEDBYINACCESSIBLE), GUARDEDBYBOTTOM))
+                    return;
+                return;
+            case NULL_LITERAL: // Don't return in the case of NULL_LITERAL since the check must be made that the LHS is @GuardedByBottom.
+            default:
+        }
+
+        boolean skipSubtypeCheck = false;
+
+        // Assigning a value with a @GuardedBy annotation to a variable with a @GuardedByInaccessible annotation is always
+        // legal. However as a precaution we verify that the locks specified in the @GuardedBy annotation are held,
+        // since this is our last chance to check anything before the @GuardedBy information is lost in the
+        // assignment to the variable annotated with @GuardedByInaccessible. See the Lock Checker manual chapter discussion
```

Comment:
We have gotten into untold trouble by not documenting behavior before implementing it.  Oftentimes the behavior was just incorrect because it had not been thought through.  Other times it was confusing.  I think the behavior should be documented rather than left undocumented.  Documentation before the fact always saves a lot more time than it costs.